docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,573 @@
+command that fetches the config of a current pod, modifies it, and creates a new
+pod. This is automating the following manual workflow:
+
+1.  <code>kubectl get -o yaml --export pod <em>pod-name</em> > pod.yaml</code>
+1.  Remove <code>status</code> from <code>pod.yaml</code> and strip all values
+    from <code>metadata</code>
+1.  Add a <code>metadata.name</code>
+1.  Modify <code>spec</code> to suit debugging needs (e.g. change
+    <code>command</code> to <code>sh</code>)
+1.  <code>kubectl create -f pod.yaml</code>
+1.  <code>kubectl attach -it pod-copy-name</code>
+
+### Related Features
+
+*Running Debug* doesn't depend on other features, but other features are
+required to realize full troubleshooting functionality.
+
+#### Shared PID Namespace
+
+Tracked in [#1615](https://issues.k8s.io/1615), a shared PID namespace became
+available in Docker 1.12 and requires plumbing through the docker drivers,
+extending the infra container to reap orphaned zombies, and [rolling
+out](https://pr.k8s.io/37404).
+
+#### Pod Level cgroup
+
+Tracked in [#26751](https://issues.k8s.io/26751), Kubernetes currently enforces
+resource constraints at a container level. A container added by *Running Debug*
+cannot allocate its own resources and must instead fit within previously
+allocated pod resources.
+
+A pod could potentially be evicted during debugging, particularly if there are
+insufficient resources to allow for the troubleshooting process, but there's not
+much we can do about this until Kubernetes adds support for vertical pod
+resource scaling ([#10782](https://issues.k8s.io/10782)).
+
+### Implementation Plan
+
+#### 1.7 Alpha Release
+
+We're targeting an alpha release in Kubernetes 1.7 that includes the following
+basic functionality:
+
+*   Support in the kubelet for attaching debug containers to a running pod
+*   A `kubectl debug` command to initiate a debug container
+*   `kubectl describe pod` will list status of debug containers running in a pod
+
+Functionality in the kubelet will be hidden behind an alpha feature flag and
+disabled by default. The following are explicitly out of scope for the 1.7 alpha
+release:
+
+*   *Copy Debug* will not be implemented
+*   `kubectl describe pod` will not report the running command, only the
+    container status
+*   Garbage collection of exitted debug containers cannot be tuned
+*   Specific integration with cluster admission controller and pod security
+    policy, though these may work by default
+
+*Running Debug* is implemented in the kubelet's generic runtime manager.
+Performing this operation with a legacy (non-CRI) runtime or a cluster without
+the alpha feature enabled will result in a not implemented error. Implementation
+will be broken into the following steps:
+
+##### Step 1: Container Type
+
+Debug containers exist within the kubelet as a `ContainerStatus` without a
+container spec. As with other `kubecontainers`, their state is persisted in the
```

Comment:
I wouldn't say we persist the state (which changes). We mostly persist the spec.